phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -0,0 +1,583 @@
+            }
+        }
+        return del;
+    }
+
+    private void addToBeVerifiedIndexRows() throws IOException {
+        for (Mutation mutation : mutations) {
+            if (mutation instanceof Put) {
+                indexKeyToDataPutMap.put(getIndexRowKey((Put)mutation), (Put)mutation);
+            }
+        }
+    }
+
+    private byte[] commitIfReady(byte[] uuidValue) throws IOException {
+        if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+            ungroupedAggregateRegionObserver.checkForRegionClosing();
+            ungroupedAggregateRegionObserver.commitBatchWithRetries(region, mutations, blockingMemstoreSize);
+            uuidValue = ServerCacheClient.generateId();
+            if (verify) {
+                addToBeVerifiedIndexRows();
+            }
+            mutations.clear();
+        }
+        return uuidValue;
+    }
+
+    private class SimpleValueGetter implements ValueGetter {
+        final ImmutableBytesWritable valuePtr = new ImmutableBytesWritable();
+        final Put put;
+        SimpleValueGetter (final Put put) {
+            this.put = put;
+        }
+        @Override
+        public ImmutableBytesWritable getLatestValue(ColumnReference ref, long ts) throws IOException {
+            List<Cell> cellList = put.get(ref.getFamily(), ref.getQualifier());
+            if (cellList == null || cellList.isEmpty()) {
+                return null;
+            }
+            Cell cell = cellList.get(0);
+            valuePtr.set(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength());
+            return valuePtr;
+        }
+
+        @Override
+        public byte[] getRowKey() {
+            return put.getRow();
+        }
+
+    }
+
+    private byte[] getIndexRowKey(final Put dataRow) throws IOException {
+        ValueGetter valueGetter = new SimpleValueGetter(dataRow);
+        byte[] builtIndexRowKey = indexMaintainer.buildRowKey(valueGetter, new ImmutableBytesWritable(dataRow.getRow()),
+                null, null, HConstants.LATEST_TIMESTAMP);
+        return builtIndexRowKey;
+    }
+
+    private boolean checkIndexRow(final byte[] indexRowKey, final Put put) throws IOException {
+        byte[] builtIndexRowKey = getIndexRowKey(put);
+        if (Bytes.compareTo(builtIndexRowKey, 0, builtIndexRowKey.length,
+                indexRowKey, 0, indexRowKey.length) != 0) {
+            return false;
+        }
+        return true;
+    }
+
+    private void logToIndexToolOutputTable(byte[] dataRowKey, byte[] indexRowKey, long dataRowTs, long indexRowTs,
+                                           String errorMsg) {
+        try (Table hTable = ServerUtil.ConnectionFactory.getConnection(ServerUtil.ConnectionType.INDEX_WRITER_CONNECTION,
+                env).getTable(TableName.valueOf(IndexTool.OUTPUT_TABLE_NAME))) {
+            byte[] rowKey = new byte[Long.BYTES];
```

Comment:
Shouldn't the row key also have the table being rebuilt? What if rebuilds are going on simultaneously on two different tables? Unlikely they'd be exact same ms, but possible. 

Also, if the row key is just a timestamp, you'll hotspot (though hopefully this table will be low volume)

Reply:
Yes, it is possible but very unlikely to get the same timestamp especially when IndexTool runs are not automated. I intentionally chose a very simple row key to make the table human readable and easy to query using the HBase shell. There will be one table row for each mismatch row. The excepted volume will be low. I suggest not to optimize this for very unlikely scenarios.

Reply:
I thought about this again. Because of the "only-verify" option where the tool reports every mismatch row instead of just one, I need to make the row key unique. So, I am thinking about including the data row key in the in the row key for the output table.